content/blog/2016/2016-06-16-parallel-test-executor-plugin.adoc

```diff
@@ -0,0 +1,191 @@
+job and splits those tests in the specified number of groups.  I can then use
+those lists to execute your tests in parallel in the manner of your choosing.
+Before I talk about the details of how it does this, let’s look at the changes
+to the pipeline:
+
+
+[source,groovy]
+----
+node { /* ...unchanged... */ }
+
+void runTests(def args) {
+  /* Request the test groupings.  Based on previous test results. */
+  /* see https://wiki.jenkins-ci.org/display/JENKINS/Parallel+Test+Executor+Plugin and demo on github
+  /* Using arbitrary parallelism of 4 and "generateInclusions" feature added in v1.8. */
+  def splits = splitTests parallelism: [$class: 'CountDrivenParallelism', size: 4], generateInclusions: true
+
+  /* Create dictionary to hold set of parallel test executions. */
+  def testGroups = [:]
+
+  for (int i = 0; i < splits.size(); i++) {
+    def split = splits[i]
+
+    /* Loop over each record in splits to prepare the testGroups that we'll run in parallel. */
+    /* Split records returned from splitTests contain { includes: boolean, list: List<String> }. */
+    /*     includes = whether list specifies tests to include (true) or tests to exclude (false). */
+    /*     list = list of tests for inclusion or exclusion. */
+    /* The list of inclusions is constructed based on results gathered from */
+    /* the previous successfully completed job. One additional record will exclude */
+    /* all known tests to run any tests not seen during the previous run.  */
+    testGroups["split${i}"] = {  // example, "split3"
+      node {
+        checkout scm
+
+        /* Clean each test node to start. */
+        mvn 'clean'
+
+        def mavenInstall = 'install -Dmaven.test.failure.ignore=true'
+
+        /* Write includesFile or excludesFile for tests.  Split record provided by splitTests. */
+        /* Tell maven to read the appropriate file. */
+        if (split.includes) {
+          writeFile file: "target/parallel-test-inclusions.txt", text: split.list.join("\n")
+          mavenInstall += " -Dsurefire.includesFile=target/parallel-test-inclusions.txt"
+        } else {
+          writeFile file: "target/parallel-test-exclusions.txt", text: split.list.join("\n")
+          mavenInstall += " -Dsurefire.excludesFile=target/parallel-test-exclusions.txt"
+        }
+
+        /* Call the maven build with tests. */
+        mvn mavenInstall
+
+        /* Archive the test results */
+        step([$class: 'JUnitResultArchiver', testResults: '**/target/surefire-reports/TEST-*.xml'])
+      }
+    }
+  }
+  parallel testGroups
+}
+
+/* Run maven */
+void mvn(def args) { /* ... */ }
+----
+
+That’s it!  The change is significant but it is all encapsulated in that one method in the pipeline.
+
+=== Great (ish) Success!
+
+Here's the results for the new pipeline with parallel test execution:
+
+image::/images/post-images/2016-06-16/serial-vs-parallel.png[Pipeline Duration Comparison, role=center]
+
+TODO: still in progress...
```

Comment:
Is this TODO for the subject of the blog post, or TODO for the blog post itself? :)